app/controllers/challenges_controller.rb

```diff
@@ -0,0 +1,40 @@
+class ChallengesController < ApplicationController
+  skip_before_action :authenticate_user!, only: [:show]
+  before_action :use_challenge_id!
+
+  def show
+    if user_signed_in? && User::Challenge.where(user: current_user, challenge_id: @challenge_id).exists?
+      send("load_data_for_#{@challenge_id}")
+      render action: @challenge_id
+    else
+      render action: "external"
+    end
+  end
+
+  def start
+    begin
+      current_user.challenges.create!(challenge_id: @challenge_id)
+    rescue ActiveRecord::RecordNotUnique
+      # Catch double clicks
+    end
+
+    redirect_to action: :show
+  end
+
+  private
+  def use_challenge_id!
+    @challenge_id = params[:id]
+    redirect_to root_path unless User::Challenge::CHALLENGES.include?(@challenge_id)
+  end
+
+  def load_data_for_12in23 # rubocop:disable Naming/VariableNumber
+    @track_counts = Solution.joins(:exercise).
+      where(
+        id: current_user.iterations.where('iterations.created_at >= ?', Date.new(2022, 12, 31)).select(:solution_id)
+      ).
+      where.not('exercises.slug': "hello-world").
+      group(:track_id).count
+    @track_counts = @track_counts.sort_by(&:second).reverse.to_h
+    @tracks = Track.where(id: @track_counts.keys).index_by(&:id)
+  end
```

Comment:
@ErikSchierboom This is the only bit that's probably worth you giving much brain energy to :)

Reply:
It's quite a slow query in production at the moment. This is likely due to the `created_at >=` clause. I added a new key on the `created_at` and `solution_id` fields on the `iterations` table, and that greatly reduced the query cost locally. Might be worth trying that in production too?

Alternatively, the query could be rewritten to:

```ruby
current_user.iterations.
  joins(:exercise).
  where('iterations.created_at >= ?', Date.new(2021, 12, 31)).
  where.not('exercises.slug': "hello-world").
  group(:track_id).select(:solution_id).distinct.count
```

This loses the solution join and has the same result. It has a slightly lower query cost (around 15% lower).

Reply:
I've rewritten this to be done mainly in Ruby which is much faster. Basically any grouping seems to be pretty slow in SQL.

Reply:
> I've rewritten this to be done mainly in Ruby which is much faster.

Not pushed yet?